### PR TITLE
ci: add caching to validation workflow

### PR DESCRIPTION
**Pull request by bdura**
## Description

This PR adds caching to the validation workflow for faster CI runs.

### Types of change

This is a CI change.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [ ] I ran all tests , and all new and existing tests passed. This includes tests marked as `external` and those 
  requiring a GPU. I did this for the following directories:
  - [ ] `tests`
  - [ ] `usage_example/tests`
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: .github/workflows/validate.yml [hunk 1]
```diff
@@ -13,11 +13,20 @@ jobs:
...
+
+      - name: Set env variables
+        run: |
+          echo "PY=$(python -VV | sha256sum | cut -d' ' -f1)" >> $GITHUB_ENV
+          echo "PRECOMMIT=$(pip list | grep pre-commit)" >> $GITHUB_ENV
+
+      - uses: actions/cache@v3
+        with:
+          path: ~/.cache/pre-commit
+          key: pre-commit|${{ env.PY }}|${{ env.PRECOMMIT }}|${{ hashFiles('.pre-commit-config.yaml') }}
```
Comment (rmitsch) [inline .github/workflows/validate.yml:29]: Can you elaborate on what's happening here?

Comment (bdura) [reply .github/workflows/validate.yml:29]: Sure. GHA's caching mechanism relies on a string key to perform cache creation and retrieval.

If we break down the `pre-commit|${{ env.PY }}|${{ env.PRECOMMIT }}|${{ hashFiles('.pre-commit-config.yaml') }}` key, it means that the cache will be invalidated every time the Python version, pre-commit version or pre-commit config changes.

Comment (bdura) [reply .github/workflows/validate.yml:29]: Then the cache works roughly like this:

1. Check the key, if it exists, retrieve the cache and restore it at the given path (here `~/.cache/pre-commit`, the default pre-commit cache folder)
2. Run other actions
3. At the end of the workflow, run the "Post Run Action" which creates the cache if there was no hit

Comment (rmitsch) [reply .github/workflows/validate.yml:29]: Thanks!

**Review by rmitsch — APPROVED** [review]

**Pull request merged**